<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        // 프로미스 선언부 => promise는 하나의 상태로 변함
    var _promise = function(param){
        return new Promise(function(resolve, reject){ // 나중에 promise 객체를 생성하기 위해 promise라는 객체를 리턴하도록 함수로 감싸는 형태 
            // 비동기를 표현하기 위해 setTimeout 함수 사용
            window.setTimeout(function(){
                // param = true일때
                if(param){
                    resolve('해결 완료')
                } else{ // param = false 일때
                    reject(Error('실패'))
                }
            }, 3000)
        })
    }

    // promise 상태 
    // pending: 아직 약속을 수행중인 상태(fulfilled 또는 reject 되기 전)
    // fulfilled: 약속이 지켜진 상태 
    // rejected: 약속이 못 지켜진 상태 
    // settled: 약속 성공 유무와 상관없이 결론이 난 상태 

    // 위의 promise 선언부를 보면 나중에 promise 객체를 생성하기 위해 promise라는 객체를 리턴하도록 함수로 감싸고 있다.
    // promise 객체만 보면 파라미터로 익명함수를 담고, 익명함수는 resolve, reject를 파라미터로 받는다 
    // 일단 new Promise로 Promise가 생성되는 직후부터 resolve, reject가 호출되기 전까지의 순간을 pending(약속이 수행중인 상태)상태라고 본다 
    //  이후 비동기 작업을 마친 뒤 결과물을 약속대로 잘 줄수 있다면 첫번째 파라미터로 주입되는 resolve 함수를 호출하고 실패했다면 두번째 파라미터인 reject 함수를
    // 호출한다는것이 promise의 주요 개념


    // Promise 실행부
    _promise(true)
    .then(function(text){
        // 성공시
        console.log(text)
    }, function(error){
        // 실패시
        console.error(error)
    })

    // then, catch란 무엇인가 
    // 실행부에는 then API, catch API가 사용된다. 먼저 then은 promise 객체에서 정상적으로 비동기 작업이 완료되었을 경우 호출된다. catch는 반대로 
    // 실패했을시 실행되는 API이다.

    // catch 예제 
    _promise(true)
    .then(JSON.parse)
    .catch(function(){
        console.log(JSON.parse)
        window.alert('체이닝 중간에 에러 발생')
    })
    .then(function(text){
        console.log(text)
    })

    // 여러 promise가 모두 완료될때 실행하려면? => promise.all 
    var promise1 = new Promise(function(resolve, reject){
        window.setTimeout(function(){
            console.log('첫번째 promise 완료')
            resolve("11111");
        }, Math.random()*2000 + 1000)
    })

    var promise2 = new Promise(function(resolve, reject){
        window.setTimeout(function(){
            console.log('두번째 promise 완료')
            resolve("22222");
        }, Math.random()*1000 + 1000)
    })

     Promise.all([promise1, promise2]).then(function (values){
         console.log('모두 완료됨', values)
     })

    //  return 하지 않고 바로 new Promise로 생성하기 
    // 맨 위에서는 new Promise를 return 해주는 형태로 사용하다가 promise1, promise2 같은 경우에는 바로 new Promise를 할당하는 형태로 사용했다
    // 차이점 
    var _promise2 = new Promise(function(resolve, reject){
        // 무엇인가 수행 

        // 50% 확률로 resolve 
        if(+new Date()%2 == 0){
            resolve("stuff work")
        } 
        else {
            reject(Error("it's broke"))
        }
    })
    // 위와 같이 선언할 경우 Promise 객체에 파라미터로 넘겨준 익명함수가 즉각 실행된다. 이후 여러차례 _promise2.then(alert) 를 호출해도 이미 한번 수행되었기
    // 때문에 계속해서 resolve 또는 reject가 수행될것이다.

    // 테스트
    _promise2.then(alert).catch(alert);
    </script>
</body>
</html>